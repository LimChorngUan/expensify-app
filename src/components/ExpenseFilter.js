import React, { useCallback } from 'react';
import { connect } from 'react-redux';

import { setTextFilter, sortByDate, sortByAmount, setStartDate, setEndDate } from '../actions/filters'

const SORT_BY_VALUE = {
  DATE: `date`,
  AMOUNT: `amount`,
}

const ExpenseFilter = ({
  filters,
  setTextFilter,
  sortByDate,
  sortByAmount,
  setStartDate,
  setEndDate,
}) => {
  const handleSelectChange = useCallback(
    (sortByValue) => {
      switch(sortByValue) {
        case SORT_BY_VALUE.DATE:
          sortByDate()
          break
        case SORT_BY_VALUE.AMOUNT:
          sortByAmount()
          break
        default:
          console.log(`nah`)
      }
    }, []
  )

  return (
    <div>
      <input
        type="text"
        value={filters.text}
        onChange={setTextFilter}
        data-test-id="text-filter-input"
      />

      <label htmlFor="sort-by">Sort By:</label>
      <select
        id="sort-by"
        value={filters.sortBy}
        onChange={(e) => handleSelectChange(e.target.value)}
        data-test-id="sort-by-selector"
      >
        <option value="date">Date</option>
        <option value="amount">Amount</option>
      </select>

      <label htmlFor="start-date">Start date</label>
      <input
        id="start-date"
        type="number"
        value={filters.startDate}
        onChange={(e) => setStartDate(e.target.value)}
        data-test-id="start-date-input"
      />

      <label htmlFor="end-date">End date</label>
      <input
        id="end-date"
        type="number"
        value={filters.endDate}
        onChange={(e) => setEndDate(e.target.value)}
        data-test-id="end-date-input"
      />
    </div>
  )
}

const mapStateToProps = state => ({
  filters: state.filters,
})

const mapDispatchToProps = dispatch => ({
  setTextFilter: e => dispatch(setTextFilter(e.target.value)),
  sortByDate: () => dispatch(sortByDate()),
  sortByAmount: () => dispatch(sortByAmount()),
  setStartDate: startDate => dispatch(setStartDate(Number(startDate))),
  setEndDate: endDate => dispatch(setEndDate(Number(endDate))),
})

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(ExpenseFilter)
export { ExpenseFilter }