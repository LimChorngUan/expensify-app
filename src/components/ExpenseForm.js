import React, { useState } from 'react'

const FORM_TYPE = {
  ADD: `add`,
  EDIT: `edit`,
}

const BUTTON_TEXT = {
  [FORM_TYPE.ADD]: `Add Expense`,
  [FORM_TYPE.EDIT]: `Edit Expense`,
}

const ExpenseForm = ({
  handleSubmit,
  expense = {},
  formType
}) => {
  const [description, setDescription] = useState(expense.description || ``)
  const [amount, setAmount] = useState(expense.amount || 0)
  const [createdAt, setCreatedAt] = useState(expense.createdAt || 0)
  const [note, setNote] = useState(expense.note || ``)

  return (
    <div>
      <form onSubmit={(e) => handleSubmit(e, {
        description,
        amount,
        createdAt,
        note,
      })}>
        <input
          type="text"
          placeholder="description"
          value={description}
          onChange={(e) => setDescription(e.target.value)}
        />
        <input
          type="number"
          placeholder="amount"
          value={amount}
          onChange={(e) => setAmount(e.target.value)}
        />
        <input
          type="number"
          placeholder="created at"
          value={createdAt}
          onChange={(e) => setCreatedAt(e.target.value)}
        />
        <textarea
          placeholder="Write your notes here (optional)"
          value={note}
          onChange={(e) => setNote(e.target.value)}
        />
        <button type="submit">{BUTTON_TEXT[formType]}</button>
      </form>
    </div>
  )
}

export default ExpenseForm