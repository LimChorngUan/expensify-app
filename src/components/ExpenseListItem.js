import React from 'react'
import { Link } from 'react-router-dom'

const ExpenseListItem = ({
  id,
  description,
  amount,
  createdAt,
}) => (
  <div>
    <Link to={`/edit/${id}`}>{description}</Link>
    <p>Amount: {amount}</p>
    <p>Create at: {createdAt}</p>
  </div>
)

export default ExpenseListItem
