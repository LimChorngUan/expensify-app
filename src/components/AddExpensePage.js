import React, { useCallback } from 'react'
import { connect } from 'react-redux'

import { addExpense } from '../actions/expenses'

import ExpenseForm from './ExpenseForm'

const AddExpensePage = ({
  addExpense,
  history
}) => {
  const handleSubmit = useCallback((e, expense) => {
    e.preventDefault()
    addExpense(expense)
    history.push(`/`)
  }, [])

  return (
    <div>
      This is from my add expense component
      <ExpenseForm handleSubmit={handleSubmit} formType="add" />
    </div>
  )
}


const mapDispatchToProps = dispatch => ({
  addExpense: expense => dispatch(addExpense(expense))
})

export default connect(
  null,
  mapDispatchToProps,
)(AddExpensePage)
