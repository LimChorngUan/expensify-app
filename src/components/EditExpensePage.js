import React, { useCallback } from 'react'
import { connect } from 'react-redux'
import _ from 'lodash'

import ExpenseForm from './ExpenseForm'
import { editExpense, removeExpense } from '../actions/expenses'

const EditExpensePage = ({
  match: { params: { id } },
  expense,
  editExpense,
  removeExpense,
  history
}) => {
  const handleSubmit = useCallback(updates => {
      editExpense(id, updates)
      history.push(`/`)
    },
    []
  )

  const handleRemoveClick = useCallback(id => {
      removeExpense(id)
      history.push(`/`)
    },
    []
  )

  return (
    <React.Fragment>
      <ExpenseForm
        expense={expense}
        formType="edit"
        handleSubmit={handleSubmit}
        data-test-id="edit-expense-form"
      />
      <button
        onClick={() => {handleRemoveClick(id)}}
        data-test-id="remove-button"
      >
        Remove
      </button>
    </React.Fragment>
  )
}

const mapStateToProps = (state, ownProps) => ({
  expense: _.find(state.expenses, item => item.id === ownProps.match.params.id)
})

const mapDispatchToProps = dispatch => ({
  editExpense: (id, updates) => dispatch(editExpense(id, updates)),
  removeExpense: id => dispatch(removeExpense({ id })),
})

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(EditExpensePage)
export { EditExpensePage }