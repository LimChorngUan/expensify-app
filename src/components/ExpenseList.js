import React from 'react';
import { connect } from 'react-redux';
import _ from 'lodash'

import filterExpense from '../selectors/expenses'

import ExpenseListItem from './ExpenseListItem'

const ExpenseList = ({
  expenses
}) => (
  <div>
    <h1>My Expenses</h1>

    {!_.size(expenses) && (
      <p>No expense</p>
    )}

    {_.size(expenses) &&
      expenses.map(item => (
        <ExpenseListItem key={item.id} {...item}></ExpenseListItem>
      ))
    }
  </div>
)

const mapStateToProps = (state) => ({
  expenses: filterExpense(state.expenses, state.filters)
})

export default connect(
  mapStateToProps
)(ExpenseList)
export { ExpenseList }