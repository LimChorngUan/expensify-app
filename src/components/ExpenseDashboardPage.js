import React from 'react';
import ExpenseList from './ExpenseList';
import ExpenseFilter from './ExpenseFilter'

const ExpenseDashboardPage = () => (
  <div>
    <ExpenseFilter />
    <ExpenseList />
  </div>
);

export default ExpenseDashboardPage;
