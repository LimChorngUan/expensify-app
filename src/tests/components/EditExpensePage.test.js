import React from 'react'
import { shallow, mount } from 'enzyme'
import toJson from 'enzyme-to-json'

import { EditExpensePage } from '../../components/EditExpensePage'
import { generateTestId } from '../utils'

describe(`Component: EditExpensePage`, () => {
  let component
  const props = {
    match: { params: { id: `A` } },
    expense: {
      description: `A`,
      amount: 10,
      note: `aaa`,
      createdAt: 1
    },
    editExpense: jest.fn(),
    removeExpense: jest.fn(),
    history: { push: jest.fn() },
  }

  beforeEach(() => {
    component = mount(<EditExpensePage {...props}/>)
  })

  test(`should render correctly`, () => {
    expect(toJson(component)).toMatchSnapshot()
  })

  test(`handleSubmit should work correctly`, () => {
    const updates = { description: `B` }

    component.find(generateTestId(`edit-expense-form`)).props().handleSubmit(updates)

    expect(component.props().editExpense).toBeCalledTimes(1)
    expect(component.props().editExpense).toBeCalledWith(`A`, { description: `B`})
    expect(component.props().history.push).toBeCalledWith(`/`)
  })

  test(`handleRemoveClick should work correctly`, () => {
    component.find(generateTestId(`remove-button`)).simulate(`click`)

    expect(component.props().removeExpense).toBeCalledTimes(1)
    expect(component.props().removeExpense).toBeCalledWith(`A`)
    expect(component.props().history.push).toBeCalledWith(`/`)
  })
})