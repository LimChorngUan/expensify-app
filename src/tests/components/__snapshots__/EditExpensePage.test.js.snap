// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: EditExpensePage should render correctly 1`] = `
<EditExpensePage
  editExpense={[MockFunction]}
  expense={
    Object {
      "amount": 10,
      "createdAt": 1,
      "description": "A",
      "note": "aaa",
    }
  }
  history={
    Object {
      "push": [MockFunction],
    }
  }
  match={
    Object {
      "params": Object {
        "id": "A",
      },
    }
  }
  removeExpense={[MockFunction]}
>
  <ExpenseForm
    data-test-id="edit-expense-form"
    expense={
      Object {
        "amount": 10,
        "createdAt": 1,
        "description": "A",
        "note": "aaa",
      }
    }
    formType="edit"
    handleSubmit={[Function]}
  >
    <div>
      <form
        data-test-id="expense-form"
        onSubmit={[Function]}
      >
        <input
          data-test-id="description-input"
          onChange={[Function]}
          placeholder="description"
          type="text"
          value="A"
        />
        <input
          data-test-id="amount-input"
          onChange={[Function]}
          placeholder="amount"
          type="number"
          value={10}
        />
        <input
          onChange={[Function]}
          placeholder="created at"
          type="number"
          value={1}
        />
        <textarea
          data-test-id="note-input"
          onChange={[Function]}
          placeholder="Write your notes here (optional)"
          value="aaa"
        />
        <button
          type="submit"
        >
          Edit Expense
        </button>
      </form>
    </div>
  </ExpenseForm>
  <button
    data-test-id="remove-button"
    onClick={[Function]}
  >
    Remove
  </button>
</EditExpensePage>
`;
